sklearn/neighbors/_sorts.pyx

```diff
@@ -0,0 +1,120 @@
+# distutils : language = c++
+
+# Some internals rely on some kinds of sorts, like KNeighborsMixin for
+# partitioning neighbors.
+#
+# The C++ std library exposes a lot of efficient sorting algorithms,
+# like nth_element, which is an efficient partial sort for KNeighborsMixin's
+# case.
+#
+# To use std::algorithm, a few fixture can be defined using Cython, mainly:
+# - Cython functions used in Cython implementations. Those call
+# - C++ functions that wraps function of std::algorithm and (sometimes) use
+# - an Comparator to state how to sort
+#
+import numpy as np
+cimport numpy as np
+from cython.parallel import prange, parallel
+
+from ._typedefs import ITYPE
+
+cdef extern from *:
+    """
+    #include <algorithm>
+
+    template<class D, class I>
+    class IndexComparator {
+    private:
+        const D *data;
+    public:
+        IndexComparator(const D *data):
+            data(data) {}
+
+        bool operator()(const I &a, const I &b) const {
+            return data[a] == data[b] ? a < b : data[a] < data[b];
+        }
+    };
+
+    template<class D, class I>
+    void intro_select_inner(
+        const D *data,
+        I *indices,
+        const I &pivot,
+        const I &n_points) {
+        IndexComparator<D, I> index_comparator(data);
+        std::nth_element(
+            indices,
+            indices + pivot,
+            indices + n_points,
+            index_comparator);
+    }
+    """
+    void intro_select_inner[D, I](
+                D *data,
+                I *indices,
+                I pivot,
+                I n_points) nogil except +
+
+
+cdef int intro_select(
+        DTYPE_t *data,
+        ITYPE_t *indices,
+        ITYPE_t pivot,
+        ITYPE_t n_points) nogil except -1:
+    """Partition indices based on their associated data.
+
+    It is essentially a partial in-place quicksort around a
+    set pivot, that is upon return, the values in indices will
+    be rearranged such that (assuming numpy-style indexing):
+
+    data[indices[:pivot]] <= data[indices[pivot]] <= data[indices[pivot:]]
+
+    Parameters
+    ----------
+    data : floating pointer
+        Pointer to a 1D array of length n_points containing floating data
+    indices : int pointer
+        Pointer to a 1D array of length n_points. This will be modified
+        in-place.
+    pivot : int
+        the index within indices around which to split the points.
+    n_points : int
+        the length of data and indices.
+    Returns
+    -------
+    status : int
+        integer exit status.  On return, the contents of indices are
+        modified as noted above.
+    """
+    intro_select_inner(
+        data,
+        indices,
+        pivot,
+        n_points)
+    return 0
+
+cpdef np.ndarray[ITYPE_t, ndim=2, mode='c'] argpartition(
+        np.ndarray[DTYPE_t, ndim=2, mode='c'] data,
```

Comment:
Wouldn't it make more sense to use use typed memory views instead of the `np.ndarray` types?

Reply:
I would like to, but if we make them `const` (which is not done yet but we might want to) we won't yet be able to used fused types (https://github.com/scikit-learn/scikit-learn/issues/10624 for reference).

[This is the current case in K-Means](https://github.com/scikit-learn/scikit-learn/blob/8c4589b23c6481f978d4cfab511f25b77a805f13/sklearn/cluster/_k_means_elkan.pyx#L7-L9).

What shall we do? Go without `const` to use memory views? 🙂

Reply:
Forget what I've said: those aren't fused types.

Reply:
6506756 introduced Cython fused-types in.